### ISSUE TITLE
Please continue developing native (older) GitHub Desktop

### ISSUE THREAD

**Issue by izbor** [desktop/desktop#4946]
Please continue developing native (older) GitHub Desktop.

Electron version sucks and will always suck as it is not native (UI sucks too). It was a good attempt, but it was a mistake and it is time to admit it and move on. Everyone makes mistakes, especially someone who does something, no shame in it, we are humans. Thank you.

**Comment by shiftkey** [comment]
@izbor thanks for the feedback. We don't currently have plans to abandon this and go back to the classic apps.

> Electron version sucks and will always suck as it is not native (UI sucks too). 

Are you able to elaborate on this at all?

**Comment by izbor** [comment]
Yes, I do. Electron version is much slower, it flickers, and I don't like UI. And as far as I know, many many people continue to use native app and prefer to use the native app. If the native app will not be usable anymore I will not be switching to electron version.

**Comment by izbor** [comment]
https://github.com/desktop/desktop/issues/2885

**Comment by shiftkey** [comment]
@izbor

> Electron version is much slower

Can you be more specific on this - slower to launch, slower to do specific tasks, which OS, etc?

> it flickers

More details about this would be great as this is the first I've heard about it - which OS, your display setup, etc

> and I don't like UI. 

Again, more specific feedback here would be helpful to direct our efforts in the future.

**Comment by izbor** [comment]
OK I installed it again to test it. I use latest macbook pro.

Actually looks not as bad as it was before. It doesn't flicker that much when you resize window or switch between changed files. You fixed the issue with master and fork folders (they previously were shown with the same name). I still don't like Repositories popup - 

1) I want to see my repos all time or at least have the ability to enable /disable it. As in original version.

2) Why all commits in bold font? 

3) Pull origin button should reask if I really want to pull - its is too easy to click it by accident.

4) I think original (macos style with buttons and icons) UI is just more intuitive. I am sorry comparing original and electron version side by side I would choose original. Tell me one reason I would want to switch to new version besides that you do not update old version anymore?

**Comment by mathieudutour** [comment]
A few reasons for you to use the new version:
- unified experience across OSs -> it's easier to teach junior when you are sure that what you see is what they see
- it's all open source so if there is something you don't like, you can contribute. I added the image comparison for example and the Desktop team merged it.
- the iteration pace is a lot faster since the team only need to implement it once instead of twice.
- it's available on Linux as well

Electron doesn't suck, it's a tradeoff. And in the end, all the reasons you mentioned don't have anything to do with Electron. Coming to an open source project and asking the maintainers to shut it down is every thing but nice, hope you can see that.

**Comment by izbor** [comment]
* unified experience across OS 
 I use macos last 7 years, it might change, but it is not a reason I would switch to Electron version right now

* it's all open source
I am just a user, I don't care. Make original version open source if it is not

* the iteration pace is a lot faster since the team only need to implement it once instead of twice
You mean you need fewer developers because you don't need separate obj-c/swift team? 

* it's available on Linux as well
I do not use Linux for development atm

**Comment by RoboMWM** [comment]
Although it's missing features from the native one, the general experience with web apps is that they always tend to consume more resources than their native counterparts, especially with Electron-wrapped apps.

**Comment by NiklasBr** [comment]
I would like to add to the list of irked users. This app still can't even do basic things like spell checking (#1597), even though it is prominently displayed as "Native" on the product page. A significant regression. (Also, "The New Native"? Really)

I also consider "[a] unified experience across OS" a negative thing. Apps should always conform to the user interface paradigms of the platform they run on. During >99% of my work days I use the same operating system, I want all my apps to be working consistently in that environment. I'm not switching between computers with different operating systems on a daily basis to use this app. Really, who does? What use case exactly motivates this line of reasoning?

I'm trying to contribute (by filing bugs and some pull request), and I want to make this app better. But I still think the older version is better by a mile.

**Comment by Studio384** [comment]
@mathieudutour None of these are reasons to use the GitHub Electron version. These are reasons to develop the GitHub app with Electron.

And especially not "unified experience across OSs". Switching to Electron for "[a] unified experience across OS" is just a lie. It is to "make it easier for developers to develop this thing", it has nothing to do with the User Experience, let's be honest about that. Apps on OSses should use the native UI, not each do their own thing. If it would be a native UI and follow the design guidelines of the OS it runs on, THAT would be "for the User Experience".

And yeah, the product page calling this thing "The new native" is, of course, bullshit. Electron apps are not native.

**Comment by shiftkey** [comment]
Please note that we shared some more context and motivations about the transition [over on the GitHub Engineering blog](https://githubengineering.com/how-four-native-developers-wrote-an-electron-app/).

**Comment by mathieudutour** [comment]
> Apps should always conform to the user interface paradigms of the platform they run on. During >99% of my work days I use the same operating system

Sure but I see GitHub Desktop as a beginner to intermediate tool, and so it’s very important that when you onboard and mentor junior people, you are sure that they see the same thing as you do. It facilitates communication. And communication is the hardest thing in a team so for me it’s a good thing. I don’t really care if the button is not exactly like in the other apps. 

I can give you some other examples of companies that did a lot of user testing and found that it’s better to have a unified experience (to a certain degree) rather than a completely native UI: Dropbox used to have a really iOS like app as well as a really Android like app. They rebuilt both app to have the same design, regardless of the platform (except for the back arrow on the nav bar) and it was a huge success for collaboration between people. 

Collaboration is the main concern when using git and GitHub, you never work in a silo, and if you do or if your entire team is on the same OS, or if you never mentor people, you should try, it’s always good to broaden your horizon :)

I for one am glad that the GH team decided to develop this app. The trade offs make sense. Good job 👏

**Comment by Studio384** [comment]
@shiftkey, yeah, read that before. And it basically just confirms what I said: it wasn't done for the user experience or to make it better for the users, it was done to make it easier for the people who develop GitHub.

And the whole "the iteration pace is a lot faster" argument is also not really the point. Nice that development goes quicker now, but the idea of the "features delivered faster" melts as snow for the sun when you first have to ship something missing half of the features, then spend years trying to catch up with what you already had. Also, does it really go quicker? Because that is not what it feels like...

**Comment by shiftkey** [comment]
I want to talk some more about these tradeoffs from our side and hopefully fill in some gaps.

Full disclosure - I've been with the team since they were called GitHub for Mac and GitHub for Windows and decided to start the journey to unifying these two products under the Desktop name.

> You mean you need fewer developers because you don't need separate obj-c/swift team?

We're a small team (currently 3 engineers and some other part-time collaborators). We tried to make it work with the two existing codebases but we struggled with the technical debt while also trying to ship new things. In some alternate reality with more resources, maybe we could have pulled this off. **But this is the hand we were dealt, and we had to look at alternative solutions to ensure Desktop's survival.**

I'll also call out that we were unable to serve a potential Linux audience for similar reasons, and #1525 shows a significant interest in this that I'm trying to support, but this wasn't even feasible before the switch to Electron.

> the general experience with web apps is that they always tend to consume more resources than their native counterparts, especially with Electron-wrapped apps.

Looking at both apps on macOS when they're open to the Desktop repository (16k commits, 60+ branches) shows that the new app is doing better than the old app in terms of total memory usage (the multi-process setup is an artefact of the Chromium architecture):

<img width="718" src="https://user-images.githubusercontent.com/359239/41620531-5e4d1d6c-73e0-11e8-8596-ddd2c5aeac8b.png">

Yes, some Electron apps are not efficient with resources. I won't name names here, but it's something we're aware of and we care about. 

Some examples:

 - #3149 is what happens when you keep the application open for many days. I need to update that issue as I think I know a solution to that issue (expiring objects that are cached forever, basically).
 - #1641 is an interesting issue about the app's behaviour when minimized - I need time to dig into Electron's internals and see whether the app is unnecessarily stealing CPU time.

So if you're noticing high memory or CPU usage with some repositories, please open an issue with details. You have the ability to [grab a timeline profile](https://github.com/desktop/desktop/blob/master/docs/contributing/timeline-profile.md) or [a heap analysis](https://github.com/desktop/desktop/issues/3149#issuecomment-379529330) to help us make things better.

> During >99% of my work days I use the same operating system, I want all my apps to be working consistently in that environment.

We want to be an example of an app that fits nicely on macOS and Windows and Linux - the styled menu bar on Windows and Linux and the upcoming dark theme are examples of this. Keep submitting these things that don't seem right so we're aware of them and can track them - doing this in the open means that others can see the improvements we land and leverage our work elsewhere.

> Also, does it really go quicker? Because that is not what it feels like...

I want to call out two features that we have in Desktop Electron today that were never in the classic apps. There's probably others, but these are important to me for reasons I'll explain at the end:

 - a vastly improved image diff experience
 - syntax highlighting supporting a multitude of languages

Here's what the classic Desktop on macOS does with a commit that changes an image:

<img width="1037" src="https://user-images.githubusercontent.com/359239/41616701-fa99d5f4-73d4-11e8-83e1-7f092823e0a2.png">

And here's what it does with a complex TypeScript file:

<img width="1239" src="https://user-images.githubusercontent.com/359239/41617156-1f4dff8c-73d6-11e8-90ca-ab217efed668.png">

Compare that to what you currently see in Desktop Electron:

<img width="1303" src="https://user-images.githubusercontent.com/359239/41616777-34d537e0-73d5-11e8-97f9-d62bd3b8af54.png">

<img width="1328" src="https://user-images.githubusercontent.com/359239/41617206-3be3a282-73d6-11e8-84e3-586d0f90e0cc.png">

These features have had significant contributions from the community, and would not have been possible to ship in a timely manner with the way we were doing things previously. I cannot stress this enough: **we're able to do things now that we never could before, even with a small team**.

Anyway, to recap:

 - we don't have plans to support the classic apps or revive them
 - we want you using Desktop Electron not because it's a new trend, but because **it's the way forward for us to ship things for our users**
 - we do want your feedback on things that we can improve the Electron app
 - ensure your feedback is stuff that we can action - **please be specific and detailed**

**Comment by NiklasBr** [comment]
> I can give you some other examples of companies that did a lot of user testing and found that it’s better to have a unified experience (to a certain degree) rather than a completely native UI

I don't think I believe you, so please go ahead. 👍

**Comment by shiftkey** [comment]
I'm going to close this out because this isn't the place for an ongoing discussion about the merit of the classic apps versus the Electron app.
